cypress/support/cypress-support/src/utils.js

```diff
@@ -672,28 +672,72 @@ global.resolveDeviceVariable = function (key) {
  *
  */
 
-class FireLog {
+class FireLog extends Function {
   constructor() {
-    if (!FireLog.instance) {
-      FireLog.instance = this;
-    }
+    // Creating the function body dynamically
+    const functionBody = `
+      return function (...args) {
+        return this.log(...args);
+      }
+    `;
+    super('...args', functionBody);
+
+    const handler = {
+      apply: function (target, thisArg, argumentsList) {
+        let message;
+        const methodName = target.name;
+        if (target.hasOwnLog) {
+          // If the method has its own logging, just apply it
+          return Reflect.apply(target, thisArg, argumentsList);
+        } else {
+          if (argumentsList.length > 2)
+            message =
+              argumentsList[argumentsList.length - 1] +
+              ' expected ' +
+              JSON.stringify(argumentsList[0]) +
+              ' to be ' +
+              JSON.stringify(argumentsList[1]);
+          else
+            message =
+              argumentsList[argumentsList.length - 1] +
+              ' Actual: ' +
+              JSON.stringify(argumentsList[0]);
+          return cy.log(message).then(() => {
+            return Reflect.apply(target, thisArg, argumentsList);
+          });
+        }
+      },
+    };
+    // Proxy for the fireLog method
+    const instanceProxy = new Proxy(this, handler);
+    const fireLogProxy = new Proxy(instanceProxy, {
+      apply: function (target, thisArg, argumentsList) {
+        const message = argumentsList[argumentsList.length - 1];
+        return cy.log(message);
+      },
+    });
 
     // Use cy.log(message) for every method in the class
-    const prototype = Object.getPrototypeOf(this);
+    const prototype = Object.getPrototypeOf(instanceProxy);
     Object.getOwnPropertyNames(prototype).forEach((method) => {
-      if (method !== 'constructor' && typeof this[method] === 'function') {
-        const originalMethod = this[method];
-        this[method] = function (...args) {
-          const message = args[args.length - 1];
-
-          return cy.log(message).then(() => {
-            return originalMethod.apply(this, args);
-          });
-        };
+      if (
+        method !== 'constructor' &&
+        method !== 'fireLog' &&
+        typeof instanceProxy[method] === 'function'
+      ) {
+        instanceProxy[method] = new Proxy(instanceProxy[method], handler);
+        if (['includes'].includes(method) || ['fail'].includes(method)) {
```

Comment:
Do we need to specify each function that has it's own custom log in this if condition?
That will not be scalable. Can this be done dynamically?

Reply:
added dynamic check